clar2wasm/src/standard/standard.wat

```diff
@@ -178,182 +178,153 @@
-        (if (i64.lt_u (local.get $product0) (local.get $product1))
-            (call $runtime-error (i32.const 0))
+        (if (i32.ge_u (local.get $lz) (i32.const 128))
+            ;; product cannot overflow if the sum of leading zeros is >= 128
+            (return (call $mul-int128 (local.get $a_lo) (local.get $a_hi) (local.get $b_lo) (local.get $b_hi)))
         )
-        ;; a3b0 + a2b1 + a1b2
-        (local.set $product0 (i64.add (local.get $product0) (local.get $product2)))
-        ;; check for carry
-        (if (i64.lt_u (local.get $product0) (local.get $product2))
+
+        (if (i32.le_u (local.get $lz) (i32.const 126))
+            ;; product will overflow if the sum of leading zeros is <= 126
             (call $runtime-error (i32.const 0))
         )
-        ;; a3b0 + a2b1 + a1b2 + a0b3
-        (local.set $product0 (i64.add (local.get $product0) (local.get $product3)))
-        ;; check for carry
-        (if (i64.lt_u (local.get $product0) (local.get $product3))
-            (call $runtime-error (i32.const 0))
+
+        ;; Other case might overflow. We compute (a * b/2) and check if result > 2**127
+        ;;    -> if yes, overflow
+        ;;    -> if not, we double the product, and add b one more time if b was odd
+        
+        ;; b / 2
+        (i64.or
+            (i64.shl (local.get $b_hi) (i64.const 63))
+            (i64.shr_u (local.get $b_lo) (i64.const 1))
         )
-        ;; check for overflow in upper 32 bits of result
-        (if (i64.ne (i64.shr_u (local.get $product0) (i64.const 32)) (i64.const 0))
+        (i64.shr_u (local.get $b_hi) (i64.const 1))
+
+        ;; a * b/2
+        (call $mul-int128 (local.get $a_lo) (local.get $a_hi))
+        (local.set $a_hi)
+        (local.set $a_lo)
+
+        ;; if result/2 > 2**127, meaning clz == 0 -> overflow
+        (if (i64.eqz (i64.clz (local.get $a_hi)))
             (call $runtime-error (i32.const 0))
         )
-        ;; result += (a3b0 + a2b1 + a1b2 + a0b3) << 96
-        (local.set $product0 (i64.shl (local.get $product0) (i64.const 32)))
-        (local.set $res_hi (i64.add (local.get $res_hi) (local.get $product0)))
-        ;; check for carry
-        (if (i64.lt_u (local.get $res_hi) (local.get $product0))
-            (call $runtime-error (i32.const 0))
+
+        ;; res *= 2, meaning res <<= 1
+        (local.set $a_hi
+            (i64.or
+                (i64.shl (local.get $a_hi) (i64.const 1))
+                (i64.shr_u (local.get $a_lo) (i64.const 63))
+            )
         )
+        (local.set $a_lo (i64.shl (local.get $a_lo) (i64.const 1)))
 
-        ;; Return the result
-        (return (local.get $res_lo) (local.get $res_hi))
+        ;; if b is odd, we add b
+        (if (i32.wrap_i64 (i64.and (local.get $b_lo) (i64.const 1)))
+            (then
+                (call $add-uint (local.get $a_lo) (local.get $a_hi) (local.get $b_lo) (local.get $b_hi))
+                (local.set $a_hi)
+                (local.set $a_lo)
```

Comment:
This setting and getting at the end here could probably be optimized away a bit, but it might sacrifice readability.

Reply:
Not sure you can use the stack that starts outside of the "if" statement inside. It's not the same scope.

Reply:
Nope, I'm wrong, it's not an issue with the scope.
But there is still an issue: the `if true` branch returns a result, but not the `if false` branch.
We could have an `id` function I guess.